test/object-store/sync/app.cpp

```diff
@@ -2829,6 +2841,98 @@ TEST_CASE("app: jwt login and metadata tests", "[sync][app]") {
     }
 }
 
+TEST_CASE("app: user interrupts integration of remote changes", "[sync][app]") {
+    std::string base_url = get_base_url();
+    const auto partition = random_string(100);
+    REQUIRE(!base_url.empty());
+
+    Schema schema1{
+        {"TopLevel",
+         {
+             {"_id", PropertyType::ObjectId | PropertyType::Nullable, true},
+             {"list_of_strings", PropertyType::Array | PropertyType::String},
+         }},
+    };
+
+    auto server_app_config = minimal_app_config(base_url, "interrupt_integration", schema1);
+    TestAppSession test_session(create_app(server_app_config));
+    auto app = test_session.app();
+
+    // Upload changes from first client.
+    {
+        SyncTestFile config(app, partition, schema1);
+        auto realm = Realm::get_shared_realm(config);
+
+        CppContext c(realm);
+        for (int i = 0; i < 10; ++i) {
+            realm->begin_transaction();
+            auto obj = Object::create(c, realm, "TopLevel",
+                                      std::any(AnyDict{{"_id", ObjectId::gen()}, {"list_of_strings", AnyVector{}}}));
+            List str_list(obj, realm->schema().find("TopLevel")->property_for_name("list_of_strings"));
+            for (int j = 0; j < 100; ++j) {
+                str_list.add(c, std::any(std::string(1024, 'a' + (j % 26))));
+            }
+            realm->commit_transaction();
+        }
+
+        wait_for_upload(*realm);
+    }
+
+    // Start second client and download the changes from first client.
+
+    // Use different table name to avoid schema conflicts when merging changes from the other client.
+    Schema schema2{
+        {"TopLevel2",
+         {
+             {"_id", PropertyType::ObjectId | PropertyType::Nullable, true},
+             {"list_of_strings", PropertyType::Array | PropertyType::String},
+         }},
+    };
+    SyncTestFile config(app, partition, schema2);
+
+    SharedRealm realm;
+    std::thread th;
+    sync::version_type user_commit_version = UINT_FAST64_MAX;
+
+    config.sync_config->on_download_message_received_hook = [&](std::weak_ptr<SyncSession>, const sync::SyncProgress&,
+                                                                int64_t, sync::DownloadBatchState batch_state,
+                                                                size_t num_changesets) {
+        CHECK(batch_state == sync::DownloadBatchState::SteadyState);
+        if (num_changesets == 0)
+            return;
+
+        // Interrupt integration of server changesets by commiting a local change.
+        auto func = [&] {
+            auto db = TestHelper::get_db(realm);
```

Comment:
Won't this thread race with the sync client? Like, can't the sync client's thread get scheduled such that it's able to integrate all the changesets before we get to line 2909 here. I admit I wasn't able to make this happen in some local testing, but this seems racy to me.

Reply:
It is racy by nature, but I haven't run into problems either. I had worked on a solution, but that wasn't bulletproof either.